llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py

```diff
@@ -0,0 +1,150 @@
+        self.correct_output_parser = correct_output_parser or CorrectOutputParser
+        self.verbose = verbose
+
+    def initialize_step(self, task: Task, **kwargs: Any) -> TaskStep:
+        """Initialize step from task."""
+        sources: List[ToolOutput] = []
+        # temporary memory for new messages
+        new_memory = ChatMemoryBuffer.from_defaults()
+
+        # put current history in new memory
+        messages = task.memory.get()
+        for message in messages:
+            new_memory.put(message)
+
+        # initialize task state
+        task_state = {
+            "sources": sources,
+            "new_memory": new_memory,
+        }
+        task.extra_state.update(task_state)
+
+        return TaskStep(
+            task_id=task.task_id,
+            step_id=str(uuid.uuid4()),
+            input=task.input,
+            step_state={"prev_reasoning": ""},
+        )
+
+    @dispatcher.span
+    async def _arun_step(
+        self,
+        step: TaskStep,
+        task: Task,
+    ) -> TaskStepOutput:
+        """Run step."""
+        if self.verbose:
+            print(f"> Running step {step.step_id} for task {task.task_id}.\n")
+
+        # run CRITIC reflection
+        num_iters = 3
+        critic_history = []
+        for i in range(num_iters):
+            correct_prompt = ""
+            critic_history.put(
+                ChatMessage(role=MessageRole.USER, content=correct_prompt)
+            )
+            response = self.llm.predict_and_call(
+                self.tools, chat_history=critic_history
+            )
+            parsed_correct_response = await self.correct_output_parser.aparse(
+                response.message.content
+            )
+            critic_history.put(
+                ChatMessage(role=MessageRole.ASSISTANT, content=parsed_correct_response)
+            )
+
+            # run critique step
+            critique_prompt = ""
+            critic_history.put(
+                ChatMessage(role=MessageRole.USER, content=critique_prompt)
+            )
+            response = self.llm.predict_and_call(
+                self.critique_tools, chat_history=critic_history
+            )
+            parsed_critique_response = await self.critique_output_parser.aparse(
+                response.message.content
+            )
+            critic_history.put(
+                ChatMessage(
+                    role=MessageRole.ASSISTANT, content=parsed_critique_response
+                )
+            )
+
+        # after CRITIC reflection and correction prepare TaskStepOutput
+        task.extra_state["new_memory"].put(
+            ChatMessage(content=parsed_correct_response), role=MessageRole.ASSISTANT
+        )
+        new_steps = [
+            TaskStep(
+                task_id=task.task_id,
+                step_id=str(uuid.uuid4()),
+                input=task.input,
+                step_state={
+                    "prev_reasoning": parsed_correct_response,
```

Comment:
I'm not sure if I'm using this correctly, but it seems like this is just some intermediate step state that I can use to carry out the reasoning loop?

Reply:
It's basically state attached to the step. So on the next step, you could access this field on the step object